### PR TITLE
Add assignments as LTI to LMS from CircuitVerse | GSoC'21 Ayan Biswas

### PR DESCRIPTION
**Pull request by ayan-biswas0412**
#### Describe the changes you have made in this PR -

1. CircuitVerse as an LTI can be added to Moodle , the key and secret is stored in the database in the ```group``` table
    - new fields added to the ```assignments``` they are ```lti_consumer_key``` and ```lti_shared_secret```
2. Teacher can link new assignments to the LMS
3. Automatic login for the following cases
    -  If the user is mentor (teacher) of the group 
4. Restricting automatic login for the following cases
   - if the user is student

#### For Teachers
image

#### For Students
image

#### Setup the LTI in LMS side
image

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 class ApplicationController < ActionController::Base
   include Pundit
   protect_from_forgery with: :exception
+  skip_before_action :verify_authenticity_token, only: :launch # for lti integration
```
Comment (satu0king) [inline app/controllers/application_controller.rb:6]: Can this be moved into LTIController?

File: app/controllers/assignments_controller.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@ class AssignmentsController < ApplicationController
   before_action :set_group
   before_action :check_access, only: %i[edit update destroy reopen]
   after_action :check_reopening_status, only: [:update]
+  after_action :allow_iframe, only: %i[show]
```
Comment (satu0king) [inline app/controllers/assignments_controller.rb:9]: rename to `allow_iframe_lti` as it is more explicit

File: app/controllers/lti_controller.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        @assignment.lti_shared_secret, # the group's lti_token
+        params
+      )
+
+      if not @provider.valid_request?(request) # checking the lti request from the lms end
+        render :launch_error, status: 401
+        return
+      end
+
+      @@launch_params=params;
```
Comment (satu0king) [inline app/controllers/lti_controller.rb:29]: Why is class variable being set?

File: app/controllers/lti_controller.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+          user_in_group = GroupMember.find_by(user_id:@user.id,group_id:@group.id) # check if the user belongs to the cv group
+
+          if user_in_group.present? # user is member of the group
+            # render the button
+            flash[:notice] = "Please open the Assigment in CircuitVerse"
+            render :open_incv
+
+          else # user is not a member of the group
+            # send the user an email
+            flash[:notice] = "Check your email for a group invitation from "+@group.name+" group, first join the group then try again."
```
[the remark this answers is not in the document]
Comment (satu0king) [reply app/controllers/lti_controller.rb:47]: Please use String Interpolation here and wherever possible - http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html

[the remark this answers is not in the document]
Comment (satu0king) [reply app/controllers/lti_controller.rb:47]: Message needs to say ask your mentor/admin to add you to the group.

File: app/controllers/simulator_controller.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@ class SimulatorController < ApplicationController
   before_action :check_view_access, only: %i[show embed get_data]
   before_action :check_edit_access, only: %i[edit update update_image]
   skip_before_action :verify_authenticity_token, only: %i[get_data create]
-  after_action :allow_iframe, only: :embed
+  after_action :allow_iframe, only: %i[show embed]
```
Comment (satu0king) [inline app/controllers/simulator_controller.rb:13]: There is a reason why we have different iframe options for embed and show. I don't think we should allow iframe for all show requests. 
We have 2 options:
1. Allow iFrame Embed for show just for LTI sessions
2. Use embed route instead of show route for LTI sessions

I think 1st one is neater. Alternatively, we can allow iFrame for all LTI sessions.

File: app/helpers/lti_helper.rb [hunk 6]
```diff
@@ -0,0 +1,2 @@
+module LtiHelper
```
Comment (satu0king) [inline app/helpers/lti_helper.rb:1]: I think you can create a function like `isLTI?` which checks the session and says whether it is an LTI session or not. Similarly, we can have a `setLTI`

File: config/public.pem [hunk 7]
```diff
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAuGhk5EoG0ujwKcqU3bUo
+hUnncDQZgoGTbMt4vXXYRRklFlU7ENPrQ0nqzfWpvJl+SQoJMNMjiYbZ/o200sID
+xBo/+AR5hUFlmcOg9q1/7cJlGQpgv0TnfbSaE85ayUd4IMDcjYjc9YPFBSsGAW4D
+/ofHxUV/N8zIoCyxdBj6cX/bnH8JmkB0alp3P/R6M27i4zn8Q/G3knt4gqUeenpd
+unI7vUtbZg7MCIEwgJrP+QlL25OiEBibTkDc6U2Eg8e6FWloWADID0kcXilLwPtd
+zkYfgkmU/7zIRmSzHXT7eMSpEEDFSbxBSVt4dxKULInrRKWAZI9kqhX2BzZAZeqY
+JQIDAQAB
+-----END PUBLIC KEY-----
```
Comment (satu0king) [inline config/public.pem:9]: This has to be loaded via an environment variable or at least removed from the version control. Also what exactly is this used for?

File: config/routes.rb [hunk 8]
```diff
@@ -92,6 +92,12 @@
     get "tags/:tag", to: "projects#get_projects", as: "tag"
   end
 
+  # lti
+  get 'lti/launch'
+  post 'lti/launch'
+  get 'lti/submitscore'
+  post 'lti/submitscore'
```
Comment (satu0king) [inline config/routes.rb:99]: I think you can use rails scopes here. 
https://devblast.com/b/rails-5-routes-scope-vs-namespace

File: app/controllers/lti_controller.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        render :launch_error, status: 401
+        return
+      end
+
+      @@launch_params=params;
+      @user = User.find_by(email: @email_from_lms) # find user by matching email with circuitverse and lms 
+
+      if @user.present? # user is present in cv 
+        if @user.id === @group.mentor_id # user is teacher
+          sign_in(@user) # passwordless sign_in the user as the authenticity is verified via lms
```
Comment (satu0king) [inline app/controllers/lti_controller.rb:34]: I think you should set LTI session `session[:isLTI]=true`  only after sign in? CC: @tachyons

File: app/controllers/lti_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:47]: ook sure @satu0king

File: app/controllers/lti_controller.rb (hunk 9 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:34]: LTI session should not depend on sign in because for student end there is no sign in , and for error cases , say like any teacher added some wrong key and secret then also the session is LTI but they are restricted to access the course

File: config/public.pem (hunk 7 above)
Comment (ayan-biswas0412) [reply config/public.pem:9]: sure, I mistakenly added this file

Comment (ayan-biswas0412) [reply config/public.pem:9]: this was generated by the opensssl command as written in SETUP.md

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (ayan-biswas0412) [reply app/controllers/application_controller.rb:6]: tried but giving some error there

File: config/public.pem (hunk 7 above)
Comment (tachyons) [reply config/public.pem:9]: This is for api authentication, no need to put this in public repo, can be gitignored

File: app/controllers/simulator_controller.rb (hunk 5 above)
Comment (ayan-biswas0412) [reply app/controllers/simulator_controller.rb:13]: sure then i will do the 1st one

File: config/public.pem (hunk 7 above)
Comment (ayan-biswas0412) [reply config/public.pem:9]: yah @tachyons I will delete it

File: app/controllers/lti_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: done

File: app/controllers/simulator_controller.rb (hunk 5 above)
Comment (ayan-biswas0412) [reply app/controllers/simulator_controller.rb]: done

File: config/public.pem (hunk 7 above)
Comment (ayan-biswas0412) [reply config/public.pem]: deleted the file

File: app/controllers/assignments_controller.rb (hunk 2 above)
Comment (ayan-biswas0412) [reply app/controllers/assignments_controller.rb]: done

File: app/controllers/lti_controller.rb (hunk 3 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: deleted now

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (ayan-biswas0412) [reply app/controllers/application_controller.rb:6]: yah sure

File: config/routes.rb (hunk 8 above)
Comment (ayan-biswas0412) [reply config/routes.rb]: done

File: app/controllers/lti_controller.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+          lms_auth_success_notice = 'Logged in as '+@email_from_lms+' via '+@lms_type+' for course '+@course_title_from_lms
+          redirect_to group_assignment_path(@group,@assignment), notice: lms_auth_success_notice # if auth_success send to group page
```
[the remark this answers is not in the document]
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:36]: done

File: app/controllers/lti_controller.rb [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        else
+          user_in_group = GroupMember.find_by(user_id:@user.id,group_id:@group.id) # check if the user belongs to the cv group
```
[the remark this answers is not in the document]
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:38]: done

File: app/helpers/lti_helper.rb (hunk 6 above)
Comment (ayan-biswas0412) [reply app/helpers/lti_helper.rb:1]: is it needed for our current functionality ?

File: app/controllers/lti_controller.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
+class LtiController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :launch # for lti integration
+  before_action :set_group_assignment, only: %i[launch]
+  before_action :set_lti_params, only: %i[launch]
+  after_action :allow_iframe_lti, only: %i[launch]
+  
+  def launch
+    session[:isLTI]=true # the lti session starting
+    require 'oauth/request_proxy/action_controller_request'
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:9]: Why was this needed ?

File: app/controllers/lti_controller.rb [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    if !@assignment.present?
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:11]: if @assignment.blank ?

File: app/controllers/lti_controller.rb [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    end
+    
+    if @group.present? # if there is a valid group based for the lti_token_key
+      @provider = IMS::LTI::ToolProvider.new(
+        params[:oauth_consumer_key], # lms_oauth_consumer_key
+        @assignment.lti_shared_secret, # the group's lti_token
+        params
+      )
+
+      if not @provider.valid_request?(request) # checking the lti request from the lms end
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:25]: Use ! Instead of not

File: app/controllers/lti_controller.rb [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        render :launch_error, status: 401
+        return
+      end
+
+      @user = User.find_by(email: @email_from_lms) # find user by matching email with circuitverse and lms 
+
+      if @user.present? # user is present in cv 
+        if @user.id === @group.mentor_id # user is teacher
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:33]: Two equal is enough

File: app/controllers/lti_controller.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+          sign_in(@user) # passwordless sign_in the user as the authenticity is verified via lms
+          lms_auth_success_notice = "Logged in as #{@email_from_lms} via #{@lms_type} for course #{@course_title_from_lms}"
+          redirect_to group_assignment_path(@group,@assignment), notice: lms_auth_success_notice # if auth_success send to group page
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:36]: Missing space after comma

File: app/controllers/lti_controller.rb (hunk 14 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: done

File: app/controllers/lti_controller.rb (hunk 16 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: done

File: app/controllers/lti_controller.rb (hunk 15 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: done

File: app/controllers/lti_controller.rb (hunk 13 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: done

File: app/controllers/lti_controller.rb (hunk 12 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: mistakenly added now removed

**Review by Shivansh2407 — APPROVED** [review]
Looks good to me @ayan-biswas0412 :tada: . Great work :+1: 
@satu0king @tachyons we can go ahead and merge this. Please let me know your views.

**Comment by tachyons** [comment]
Can be merged once tests are added 👍

File: app/controllers/lti_controller.rb (hunk 12 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb]: image
else this error is happening if I remove this

**Comment by ayan-biswas0412** [comment]
> Can be merged once tests are added 👍
> […](#)

added tests @tachyons please have a review

File: app/assets/javascripts/lti.coffee [hunk 17]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (tachyons) [inline app/assets/javascripts/lti.coffee:1]: remove this file

File: app/assets/stylesheets/lti.scss [hunk 18]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the lti controller here.
```
Comment (tachyons) [inline app/assets/stylesheets/lti.scss:1]: remove this file

File: app/controllers/assignments_controller.rb [hunk 19]
```diff
@@ -108,6 +109,12 @@ def destroy
     end
   end
 
+  def allow_iframe_lti
+    return unless session[:isLTI]
+
+    response.headers.except! "X-Frame-Options"
```
Comment (tachyons) [inline app/controllers/assignments_controller.rb:115]: Can we use     `response.headers["X-FRAME-OPTIONS"] = "ALLOW-FROM http://example.com"` to white list just the moodle url if we can get moodle url in the session ?

File: app/controllers/lti_controller.rb [hunk 20]
```diff
@@ -0,0 +1,81 @@
+class LtiController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :launch # for lti integration
+  before_action :set_group_assignment, only: %i[launch]
+  before_action :set_lti_params, only: %i[launch]
+  after_action :allow_iframe_lti, only: %i[launch]
+  
+  def launch
+    session[:isLTI]=true # the lti session starting
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:8]: Use snake case, `is_lti`

File: app/controllers/lti_controller.rb (hunk 12 above)
Comment (tachyons) [reply app/controllers/lti_controller.rb:9]: Then move this to top of the file

File: app/controllers/lti_controller.rb [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+      if @user.present? # user is present in cv 
+        if @user.id == @group.mentor_id # user is teacher
+          sign_in(@user) # passwordless sign_in the user as the authenticity is verified via lms
+          lms_auth_success_notice = t("lti.launch.notice_lms_auth_success_teacher", email_from_lms: @email_from_lms, lms_type: @lms_type, course_title_from_lms: @course_title_from_lms)
+          redirect_to group_assignment_path(@group, @assignment), notice: lms_auth_success_notice # if auth_success send to group page
+        else
+          user_in_group = GroupMember.find_by(user_id:@user.id,group_id:@group.id) # check if the user belongs to the cv group
+
+          if user_in_group.present? # user is member of the group
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:40]: should be something like @group.members.exists?(user: @user)

File: app/controllers/lti_controller.rb [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    else # there is no valid group present for the lti_consumer_key
+      flash[:notice] = t("lti.launch.notice_invalid_group")
+      render :launch_error, status: 400
+    end
+  end
+
+  def allow_iframe_lti
+    return unless session[:isLTI]
+    
+    response.headers.except! "X-Frame-Options"
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:64]: As as before, try if we can whitelist specific domain instead of all

File: app/controllers/lti_controller.rb [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  private
+
+    def set_group_assignment # query db and check lms_oauth_consumer_key is equal to which assignment and find the group also
+      @assignment = Assignment.find_by(lti_consumer_key: params[:oauth_consumer_key])
+      if @assignment.present?
+        @group = Group.find_by(id: @assignment.group_id)
```
Comment (tachyons) [inline app/controllers/lti_controller.rb:72]:
```suggestion
        @group =@assignment.group
```

File: app/views/lti/launch.html.erb [hunk 24]
```diff
@@ -0,0 +1,2 @@
+<h1>Lti#launch</h1>
```
Comment (tachyons) [inline app/views/lti/launch.html.erb:1]: Remove if not in use

File: spec/controllers/lti_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,21 @@
+require "rails_helper"
+
+describe LtiController, type: :request do
```
Comment (tachyons) [inline spec/controllers/lti_controller_spec.rb:3]: should be in spec/requests folder

Comment (ayan-biswas0412) [reply spec/controllers/lti_controller_spec.rb:3]: entire code of ```spec/controllers/lti_controller_spec.rb```  should be in ```spec/requests``` ??

File: app/controllers/lti_controller.rb (hunk 21 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:40]: This error is happening
image

File: app/controllers/lti_controller.rb (hunk 12 above)
Comment (ayan-biswas0412) [reply app/controllers/lti_controller.rb:9]: done

**Review by tachyons — APPROVED** [review]

File: app/controllers/assignments_controller.rb [hunk 26]
```diff
@@ -108,6 +109,12 @@ def destroy
     end
   end
 
+  def allow_iframe_lti
+    return unless session[:is_lti]
+  
+    response.headers["X-FRAME-OPTIONS"] = "ALLOW-FROM #{session[:lms_domain]}"
```
Comment (satu0king) [inline app/controllers/assignments_controller.rb:115]: @ayan-biswas0412 , check this once, might apply
https://stackoverflow.com/questions/10205192/x-frame-options-allow-from-multiple-domains

File: app/controllers/simulator_controller.rb [hunk 27]
```diff
@@ -94,6 +95,12 @@ def verilog_cv
     render json: response.to_s, status: response.code
   end
 
+  def allow_iframe_lti
```
Comment (satu0king) [inline app/controllers/simulator_controller.rb:98]: Seems to be repeated in 3 places, can it be refactored?

**Review by satu0king — APPROVED** [review]

File: app/controllers/simulator_controller.rb (hunk 27 above)
Comment (ayan-biswas0412) [reply app/controllers/simulator_controller.rb:98]: will try to refractor in next PR
